### DISCUSSION TITLE
Mixins, Classes, TS

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#8419] [General]
I have been thinking how to migrate complex modules like Text/Canvas/Object and their mixins.
The difficulty is that the base class and the mixins are tangled together, cross referencing each others methods.
Some mixins are well defined and can be made as subclasses like @asturur did with ObjectGeometry (super class) and ObjectInteractivity (subclass).
Other mixins are tangled...
I think of using abstract classes to bridge the tangle by defining an abstract super class that all derived classes will extend and implement partially. Then we chain them together (preferably by extending) and produce the final mixed class.

Using abstract classes is correct IMO since this mixins are abstract and so are the objects they mix since none are standalone.

```ts
abstract class Base {
    abstract foo(): number
    abstract bar(): number
}

abstract class A extends Base {
    foo() {
        return 1 + this.bar()
    }
}

abstract class B extends A {
    bar(){
        return 5
    }
}

class C extends B {

}

console.log(new C().bar(), new C().foo())
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class Base {
}
class A extends Base {
    foo() {
        return 1 + this.bar();
    }
}
class B extends A {
    bar() {
        return 5;
    }
}
class C extends B {
}
console.log(new C().bar(), new C().foo());

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/IYIwzgLgTsDGEAJYBthjAgQmgpgg3gLABQCZCokM8CAZgPb0AUAlAFwIB2ArgLYg4oJchXDQ4iEMCisOPfoJIBfEiUriaKNBgCCCHAA8IOTgBMM2MHiKlyDZiwLCR5KDgjconBAEYEAagQIAAsASzAAOikZFmcyFWIEtTFqRC10LH0jE3MEPRsRaNYClzI3Dy8EAFY4hASk4nSMAGEs4zMLJ2JlVUb6TjB6ZBwI5HoAcyZOHAB3BGbWKOlWABouWfnF+1YWIA)
      

---

Here you can see A and B are astrange to each other, C extends A and is mixed with B.
TS doesn't approve, JS does.

```ts
class A {
    foo() {
        return 1
    }
}

abstract class B {
    abstract foo():number
    bar(){
        return 5 + this.foo()
    }
}

function applyMixins<T, U>(derivedCtor: T, constructors: U[]) {
  constructors.forEach((baseCtor) => {
    Object.getOwnPropertyNames(baseCtor.prototype).forEach((name) => {
      Object.defineProperty(
        derivedCtor.prototype,
        name,
        Object.getOwnPropertyDescriptor(baseCtor.prototype, name) ||
          Object.create(null)
      );
    });
  });
  return derivedCtor as U & T
}

const Mixed = applyMixins(class C extends A{},[B]);

console.log(new Mixed().bar(), new Mixed().foo())
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class A {
    foo() {
        return 1;
    }
}
class B {
    bar() {
        return 5 + this.foo();
    }
}
function applyMixins(derivedCtor, constructors) {
    constructors.forEach((baseCtor) => {
        Object.getOwnPropertyNames(baseCtor.prototype).forEach((name) => {
            Object.defineProperty(derivedCtor.prototype, name, Object.getOwnPropertyDescriptor(baseCtor.prototype, name) ||
                Object.create(null));
        });
    });
    return derivedCtor;
}
const Mixed = applyMixins(class C extends A {
}, [B]);
console.log(new Mixed().bar(), new Mixed().foo());

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8CwAoa7oDMD22AUAlIqhqdAE4CmALgK7kB20AjCRgL6qcqpgBGEauTDBq0UJBgAhYmgz9Bw0VlyEAXA1oBbPpXJt0fMOULI5ZCjXpMArNADU0agAsAlhAB0OfAQPRu3KiYtAyirthMYAAOUSAAngCyrgAergwQADwAKgA00ACqAHx4ACZ6rgBulCUAwtTY5GrQueIRirSiDRBN+QDaALpEZujAbUId9eSeOOQAoiLOeHhGEJR1DUQAvIWypADyfABWlKIeAOY0ewDuDAAK5NhRetRxAHJgWpQQy5Brkx5RB71F5PAheBrzYCLPAMD6ULY7YZkA7HU5lTBpSj3R7POJ4PykMrkSrVdbkAFA7Agyg5AkYWGfWnmZFHE7Uc6XG7Yp7kF4AES+wGJUUmP1WZIpVKpcSeeQZ8OgAB9FXT9qzTkLKGBqJQYbQQCBfMz0AQANx+dhmtiW81yKh0RjQIkk2qTaCQArQABkzS4qFQo3SYiSyWq0E27pi8RDaW+Eig0Bq0EoyR1DBKMFgCHYOV6UkGtoDbWwIEoHhA2DOMMoV2gIeqhA8RhMBDlNbrKQbYO8hAIQA)

**Comment by ShaMan123** [comment]
I will try migrating Text with abstract class

**Reply by ShaMan123** [reply]
#8420

**Comment by ShaMan123** [comment]
@asturur @melchiar

**Comment by thuytv-gl** [comment]
Hi @ShaMan123,
I cloned and build fabric yesterday (after this migration), and now keys events are not working.
Is it happening on your side.

----
## What did I do
1. Fork master
2. Clone the fork
3. Install dependencies with `npm i --include=dev`
4. Link and build with `npm link && npm run build -- -f -w`
5. Create a sandbox with vite/svelte
6. Install the link with `npm link --save fabric`
7. Run the sandbox with few textboxes
8. Try keys events: ctrl+c, ctrl+a, arrow keys --> nothing works

**Reply by ShaMan123** [reply]
You should have opened an issue
Fixed #8457

**Reply by thuytv-gl** [reply]
thank you and noted!

**Comment by asturur** [comment]
The issue with abstract classes is that in JS they are just classes.
Between extending classes or extending abstract classes there is no difference, we are still creating nested boxes.
The only bad thing of creating nested boxes is that then is impossible to remember who owns a method. You can forget to do Object.prototype.functionX because is somewhere between N layers.

It would be interesting if mixin can create flat classes like the old fabricJS in which mixins were just a developer driven concern of splitting files and nothing more.

I think the flat classes are also, let's say, slower because are crate dynamically or at runtime.

**Reply by ShaMan123** [reply]
Wait... if you subclass you can call subclass.prototype.methodOfSuperClass, am I right?
